src/val.rs

```diff
@@ -0,0 +1,128 @@
+use std::rc::Rc;
+
+// Design notes:
+// =============
+//
+// The "list of strings" data structure can be represented in three different reasonable ways:
+//
+// 1. As a vec of strings (currently implemented).
+// This is simple and easy to debug, but makes zero-copy stuff difficult.
+// At least can make copying cheap with an Rc<T>.
+//
+// 2. As a pre-joined string + vec of indices.
+// This allows a `&str` ref to the pre-joined string to be acquired,
+// at the cost of bounds checking when the parts are accessed.
+// Would probably be the most compact and efficient implementation.
+//
+// 3. As a value that contains both a pre-joined variant and list variant.
+// This trades memory for the ability to borrow refs to both variants.
+//
+// The current implementation focuses on simplicity,
+// but tries to preserve flexibility for future optimizations
+// by mostly only dealing with owned data at the API boundaries (no borrowed views).
+
+/// The internal data type of Just, an immutable list of strings.
+///
+/// It can be viewed as a "joined string", or as the individual "parts".
+#[derive(Debug, Clone, Default)]
+pub(crate) struct Val {
+  parts: Rc<[Box<str>]>,
```

Comment:
`Rc<[Box<str>]>` is a little exotic. Is the advantage that extra capacity is freed when compared to an `Rc<Vec<String>>`?

I'd be in favor of using `Rc<Vec<String>>`, just because it's easier to read, and switching to `Rc<[Box<str>]` if we really need to.